sycl/include/CL/sycl/ONEAPI/reduction.hpp

```diff
@@ -225,6 +225,27 @@ struct known_identity_impl<BinaryOperation, AccumulatorT,
           : std::numeric_limits<AccumulatorT>::lowest();
 };
 
+#ifdef __SYCL_REDUCER_OP_EQ_CHECK_TRAIT
+#error "__SYCL_REDUCER_OP_EQ_CHECK_TRAIT must not be defined"
+#endif
+
+#define __SYCL_REDUCER_OP_EQ_CHECK_TRAIT(OpName, Op)                           \
+  template <typename, typename = void>                                         \
+  struct HasSameTypeArg##OpName##Eq : public std::false_type {};               \
+  template <typename T>                                                        \
+  struct HasSameTypeArg##OpName##Eq<                                           \
+      T, std::enable_if_t<std::is_same<                                        \
+             decltype(static_cast<T &(T::*)(const T &)>(&T::operator+=)),      \
```

Comment:
Why is this always `+=`?

Reply:
That is an error that I fixed in experimental branch, but forgot to fix here: 
need to replace "&T::operator+=" with "&T::operator Op="

Reply:
I am already not so sure if this check for existence of += is needed at all now.

What if user passes some 'binary_operation' to reduction AND also defines "T &operator+=(const T&)".
Should the reduction implementation run T::operator+= for the kernel statement: 'Sum += Data[i]' or it should rather ignore operator+= and handle it as 'Sum = binary_operation(Sum, Data[i])' ?

Reply:
You're right.  I don't think we want to call a user-defined `+=` operator at all, and should always forward to `Sum = binary_operation(Sum, Data[i])`.  In the case of a predefined operator like `Sum = std::plus<>(Sum, Data[i])` it will call the user-defined operator for `+` (because of how `std::plus` is defined).  Calling `+=` in some cases and `+` in others may lead to confusing error messages.

That suggests to me that it would be enough to just use `enable_if_t<IsReduPlus<_T, BinaryOperation>::value,           reducer &>` as the test on the operator.  Since that would make the definition of the operators in the base class and identity specialization identical, we'd only need the operators in the base class.

Reply:
BTW, excuse me if this question is naïve, but wouldn't it be better to to define reducer::operator+= as a method returning void? The reasons for that I suppose are the same as for the method reducer::combine() returning void now, i.e. because the result of operator+= cannot be used in any other context (in LHS or RHS of any expression), right?

Reply:
Not a naïve question at all!  I think this is actually a bug (or oversight) in the original reduction proposal.  The SYCL 2020 version of these operators return `void` as you suggest: https://www.khronos.org/registry/SYCL/specs/sycl-2020/html/sycl-2020.html#table.operators.reducer

Since we're modifying these operators in this PR already, changing the return type to `void` looks like a good idea.  One step closer to SYCL 2020 reductions!

Reply:
> Since that would make the definition of the operators in the base class and identity specialization identical, we'd only need the operators in the base class.

John, In your comment it look like your intent is to separate some methods into a base class.

I don't see how to do that gracefully and without using virtual methods.
The biggest difficulty is that the generic class cannot be a base class because it has 2 additional fields,
that we clearly don't want having in the child class (specialized class now).

If you know how to refactor the reducer class, please give advice, or alternatively try doing such modification in a separate change-set.

Reply:
Ah, I see.  Sorry, I misread the code -- I thought that the known identity specialization already inherited from a base class.